### PR TITLE
SCC-3986: Node20

### PR DESCRIPTION
**Pull request by nonword**
Update app to use Node 20.

 - Added qa and prod encrypted config to `./config/` and updated app to decrypt it on startup
 - Updated all dependencies (except vulnerable sub-dependencies of other nypl modules, for which I created tickets)
 - Removed many vulnerable dependencies that could be removed with minor app changes (including removing some unused data files)
 - Removed invalidated sections of README
 - Rebuilt many fixtures
 - Linting
 - GH actions for testing and deploying from the Node 20 QA release branch
 - Removed .travis because gh-actions covers things (and travis is failing `nvm i` for Node20 for some reason)

### CODE REVIEW

**Comment by charmingduchess** [comment]
Hey Paul! I'm going to take a closer look at this later, but I'm wondering for starters if we can git ignore the fixtures. I can sort of see why we'd want them committed, but I imagine the tests should theoretically pass whether the fixtures are exactly the same or not.

File: README.md [hunk 1]
```diff
@@ -1,51 +1,42 @@
...
+For local development, it's easiest to just use local node binaries:
 
 ```
-nvm use
+nvm use; npm i
+nvm use; ENV=qa npm start
 ```
 
-Install dependencies:
+Note that when developing locally, you may need to [add your IP to the access control policy of the relevant ES domain](https://github.com/NYPL/aws/blob/b5c0af0ec8357af9a645d8b47a5dbb0090966071/common/elasticsearch.md#2-make-the-domain-public-restrict-by-ip).
```
Comment (charmingduchess) [inline README.md:16]: Can this include a note about what the error looks like when you don't have permissions  eg `{"Message":"User: anonymous is not authorized to perform: es:ESHttpGet"}`? I remember being confused by this a lot when I started working with this repo.

Comment (nonword) [reply README.md]: I added more detail to this documentation - and better app handling so that the specific error is clearer in app logs

**Review by charmingduchess — COMMENTED** [review]
I think I've looked at every file that isn't a fixture, and looks good to me. I have some clarifying questions on some code removals and changes before I'd like to approve.

File: lib/es-client.js [hunk 2]
```diff
@@ -66,7 +84,7 @@ clientWrapper.nonRecapItemStatusAggregation = (bnum) => {
 
   return clientWrapper.search(esQuery)
     .then((resp) => {
-      return deepValue(resp, 'aggregations.statuses.nonrecap_statuses.nonrecap_status_buckets.buckets')
+      return deepValue(resp, 'body.aggregations.statuses.nonrecap_statuses.nonrecap_status_buckets.buckets')
```
[the remark this answers is not in the document]
Comment (nonword) [reply lib/es-client.js:96]: The newer Es client returns an object that nests the response under `body`

File: lib/jsonld_serializers.js [hunk 3]
```diff
@@ -149,14 +146,12 @@ class JsonLdItemSerializer extends JsonLdSerializer {
...
-    var label = str.split(PACK_DELIM)[1]
+  const parse = (str) => {
+    const urn = str.split(PACK_DELIM)[0]
+    const label = str.split(PACK_DELIM)[1]
 
-    var urnParts = urn.split(':')
-    var type = { terms: 'nypl:Term', agents: 'nypl:Agent' }[urnParts[0]]
-    var id = urnParts[1]
-    id = id.match(/^\d+$/) ? parseInt(id) : id
+    const urnParts = urn.split(':')
```
[the remark this answers is not in the document]
Comment (nonword) [reply lib/jsonld_serializers.js:153]: 🚰

File: lib/jsonld_serializers.js [hunk 4]
```diff
@@ -213,24 +208,13 @@ class ResourceSerializer extends JsonLdItemSerializer {
...
   }
 
   statements () {
-    var stmts = JsonLdItemSerializer.prototype.statements.call(this)
-
-    if (this.body.parentUri) stmts.memberOf = R.flatten([util.eachValue(this.body.parentUri, (id) => ({ '@type': 'nypl:Resource', '@id': `res:${id}` }))])
+    const stmts = JsonLdItemSerializer.prototype.statements.call(this)
 
-    // Parse all contributor_(aut|ill|...) statements:
-    Object.keys(this.body).forEach((field) => {
```
[the remark this answers is not in the document]
Comment (nonword) [reply lib/jsonld_serializers.js:225]: This is cruft from the original "registry" prototype that discovery-api evolved from. There used to be properties representing specific roles held by contributors. We longer do that (i.e. there are no fields matching `/contributor_\w{3}_packed/` in the ES index).

Part of this work was updating dependencies to address vulnerabilities and simplify things. One of the stale dependencies was `nypl-registry/utils-lexicon`. In order to remove it, I had to remove all references to it in the code. This was one of those references.

File: lib/jsonld_serializers.js [hunk 5]
```diff
@@ -261,7 +245,8 @@ class ResourceSerializer extends JsonLdItemSerializer {
     // Check both numItemsTotal (the property written by RCI) and numItems
     // (written by legacy indexing process). Note that very old bibs may have
     // neither
-    const numItems = this.body.numItemsTotal && this.body.numItemsTotal[0] || this.body.numItems && this.body.numItems[0]
+    const numItems = (this.body.numItemsTotal && this.body.numItemsTotal[0]) ||
```
[the remark this answers is not in the document]
Comment (nonword) [reply lib/jsonld_serializers.js:248]: Totally agree. I'm trying to maintain as much app-code parity as possible until after we launch this Node20 work. If we have to update the main branch, we'll have an easier time folding those changes into this branch if fewer things have changed. There's a ton of fun refactoring and cleanup we can do after.

File: lib/logger.js [hunk 6]
```diff
@@ -1,47 +1,23 @@
...
-  colorize: true
-}))
-
-const logger = new winston.Logger({
-  transports: loggerTransports,
-  exitOnError: false
+const logger = winston.createLogger({
+  level: process.env.LOG_LEVEL || 'info',
+  transports: [
+    new winston.transports.Console({ format })
```
[the remark this answers is not in the document]
Comment (nonword) [reply lib/logger.js:15]: Defined starting on line 4 - It is a `winston.format.json()` by default (i.e. log to JSON) or a special colorized plaintext logging format (much more readable for local development)

File: lib/load-config.js
[the remark this answers is not in the document]
Comment (nonword) [reply lib/load-config.js:1]: Yeah py-utils roughly [does a similar thing](https://github.com/NYPL/python-utils/blob/main/src/nypl_py_utils/functions/config_helper.py) - the `load_env_file` function is used to load and run all necessary decryption ahead of instantiating any clients that use that config

File: lib/util.js [hunk 7]
```diff
@@ -50,43 +42,24 @@ exports.context = (which) => {
...
         data = JSON.parse(data)
-        __file_cache[cacheKey] = data
+        __fileCache[cacheKey] = data
         resolve(data)
       }
     })
   })
 }
 
-var __file_cache = {}
```
[the remark this answers is not in the document]
Comment (nonword) [reply lib/util.js:70]: This was part of dependency cleanup. There was a critical vulnerability in `fast-csv`, which was used by some of these util functions, but on inspection we weren't using any of the util functions.

**Review by charmingduchess — APPROVED** [review]
A small step for man, a giant leap for LSP!

**Comment by nonword** [comment]
This has been approved, but we're not ready to merge this to `main` yet. This node20 work now lives in release branch `qa-node20` for testing.

**Pull request closed without merging**